src/_pytest/cacheprovider.py

```diff
@@ -203,17 +205,35 @@ def set(self, key: str, value: object) -> None:
             with f:
                 f.write(data)
 
-    def _ensure_supporting_files(self) -> None:
-        """Create supporting files in the cache dir that are not really part of the cache."""
-        readme_path = self._cachedir / "README.md"
-        readme_path.write_text(README_CONTENT, encoding="UTF-8")
+    def _ensure_cache_dir_and_supporting_files(self) -> None:
+        """Create the cache dir and its supporting files."""
+        if self._cachedir.is_dir():
+            return
 
-        gitignore_path = self._cachedir.joinpath(".gitignore")
-        msg = "# Created by pytest automatically.\n*\n"
-        gitignore_path.write_text(msg, encoding="UTF-8")
+        files: Iterable[Tuple[str, Union[str, bytes]]] = (
+            ("README.md", README_CONTENT),
+            (".gitignore", "# Created by pytest automatically.\n*\n"),
+            ("CACHEDIR.TAG", CACHEDIR_TAG_CONTENT),
+        )
 
-        cachedir_tag_path = self._cachedir.joinpath("CACHEDIR.TAG")
-        cachedir_tag_path.write_bytes(CACHEDIR_TAG_CONTENT)
+        with tempfile.TemporaryDirectory() as d:
+            for file, content in files:
+                file = os.path.join(d, file)
+                if isinstance(content, str):
+                    with open(file, "xt", encoding="UTF-8") as f:
+                        f.write(content)
+                elif isinstance(content, bytes):
+                    with open(file, "xb") as f:
+                        f.write(content)
+                else:
+                    assert_never(content)
+            shutil.move(d, self._cachedir)
```

Comment:
If it's alright with you, I like the defensive style that's here now.

Reply:
Sorry, what's defensive about it, can you clarify?

Reply:
It will raise if somehow there's a file at the destination. Obviously this isn't likely today, but this is more resistant to refactoring that might loosen surrounding guarantees. We probably never want to be overwriting here.

Reply:
> It will raise if somehow there's a file at the destination. 

Ahh OK due to the `x`/`xt` flags, sorry. Indeed, but we can do similar:

```python
        with tempfile.TemporaryDirectory() as d:
            path = Path(d)
            with open(path.joinpath("README.md"), "xt", encoding="UTF-8") as f:
                f.write_text(README_CONTENT)
            with open(path.joinpath(".gitignore"), "xt", encoding="UTF-8") as f:
                f.write_text("# Created by pytest automatically.\n*\n")
            with open(path.joinpath("CACHEDIR.TAG"), "xb") as f:
                f.write_bytes(CACHEDIR_TAG_CONTENT)        
            shutil.move(path, self._cachedir)
```

My point is that it is more straightforward to just perform the file writing directly, rather than creating a list and iterate over it, with if/else switches based on the data type.